File: Data Structures & Algorithms/house-robber/submission-1.cpp
class Solution {
public:
    map<int, int> cache;
    int rob(vector<int>& nums) {
        return dp(nums, 0);
    }

    int dp(vector<int>& nums, int start) {
        if (start >= nums.size()) {
            return 0;
        }
        if (cache.find(start) != cache.end()) {
            return cache[start];
        }

        int sum= max(nums[start] + dp(nums, start+2), dp(nums, start+1));
        cache[start] = sum;
        return sum;
    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-1.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if len(nums) < 3:
            return []
        nums.sort()
        res = set()
        for i in range(len(nums)):
            rem = -nums[i]
            start = i+1
            end = len(nums) - 1
            print (f"i : {i}, start :{start}, end :{end}")
            while start < end:
                sum = nums[start] + nums[end]
                if rem == sum:
                    res.add((nums[i], nums[start], nums[end]))
                    start += 1
                    end -= 1
                elif sum < rem:
                    start += 1
                else :
                    end -= 1
        return list(res)
